### ISSUE TITLE
[url_launcher] `'UIApplication' does not conform to protocol 'Launcher'` with Xcode16 beta3

### ISSUE THREAD

**Issue by remonh87** [flutter/flutter#151467]
### What package does this bug report belong to?

url_launcher

### What target platforms are you seeing this bug on?

iOS

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
  url_launcher_ios:
    dependency: transitive
    description:
      name: url_launcher_ios
      sha256: "7068716403343f6ba4969b4173cbf3b84fc768042124bc2c011e5d782b24fe89"
      url: "https://pub.dev"
    source: hosted
    version: "6.3.0"```

</details>


### Steps to reproduce

1. create flutter project
2. add url_launcher package
3. build it for an io18 simulator using xcode16.0 beta3

### Expected results

It builds correctly

### Actual results

It fails to build with error: 
```
  /Users/user/.pub-cache/hosted/pub.dev/url_launcher_ios-6.3.0/ios/url_launcher_ios/Sources/url_launcher_ios/Launcher.swift:22:1: error: type 'UIApplication' does not conform to protocol 'Launcher'
    extension UIApplication: Launcher {}
    ^
    UIKit.UIApplication:24:15: note: candidate has non-matching type '(URL, [UIApplication.OpenExternalURLOptionsKey : Any], (@MainActor @Sendable (Bool) -> Void)?) -> Void'
        open func open(_ url: URL, options: [UIApplication.OpenExternalURLOptionsKey : Any] = [:], completionHandler completion: (@MainActor @Sendable (Bool) -> Void)? = nil)
```

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => launchUrl(Uri.parse('www.google.com')),
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.5 23F79 darwin-arm64, locale en-GB)
    • Flutter version 3.22.2 on channel stable at /Users/user/fvm/versions/3.22.2
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (5 weeks ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/user/Library/Android/sdk
    • Platform android-35, build-tools 34.0.0
    • ANDROID_HOME = /Users/user/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode-beta.app/Contents/Developer
    • Build 16A5202i
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] VS Code (version 1.91.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.92.0

[✓] Connected device (5 available)
    • iPhone 15 (mobile)              • FD673D06-5760-4BDE-978B-D9D75D048513 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-18-0 (simulator)
    • macOS (desktop)                 • macos                                • darwin-arm64   • macOS 14.5 23F79 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad                • darwin         • macOS 14.5 23F79 darwin-arm64
    • Chrome (web)                    • chrome                               • web-javascript • Google Chrome 126.0.6478.127

[✓] Network resources
    • All expected network resources are available.```

</details>

**Comment by remonh87** [comment]
The url template looks a bit broken here but to reproduce it is:

1. create a. flutter basic app
2. Add url_launcher using `flutter pub  add url_launcher`
3. compile using xcode16 beta3 version on an iOS 18 simulator

**Comment by stuartmorgan** [comment]
It looks like either the Obj-C->Swift generation changed slightly, or conformance checks are stricter now. If it's the former we may need to conditionalize the exact declaration of the protocol method by SDK version.

**Comment by eidolonFIRE** [comment]
I am also hitting this:
```
xcode 16 beta 3
iOS 17.5.1
url_launcher 6.3.0
```

**Comment by jmagman** [comment]
This built:
```diff
-    completionHandler completion: ((Bool) -> Void)?)
+    completionHandler completion: (@MainActor @Sendable (Bool) -> Void)?)
```
https://github.com/flutter/packages/blob/01527175e38e7ecd9ec180a1f0781580adb82dd5/packages/url_launcher/url_launcher_ios/ios/url_launcher_ios/Sources/url_launcher_ios/Launcher.swift#L18

**Comment by stuartmorgan** [comment]
Thanks, I didn't have the beta install to experiment.

We just need to set the Swift version to 5.5 in the podspec to use that, right?

**Comment by jmagman** [comment]
@hellohuanlin can you take a look at this one? url_launcher should build on Xcode 14 (minimum enforced in the tool, includes Swift 5.7) as well as Xcode 16 beta 3.

**Comment by daironigr** [comment]
Any solution for this?

**Comment by hellohuanlin** [comment]
> Any solution for this?

I am working on it. I am not able to repro it on Xcode beta 1, which means Apple is likely still hashing things out in enforcing the concurrency checks. I am downloading beta 3 now and see how it goes.

**Comment by daironigr** [comment]
> > Any solution for this?
> 
> I am working on it. I am not able to repro it on Xcode beta 1, which means Apple is likely still hashing things out in enforcing the concurrency checks. I am downloading beta 3 now and see how it goes.

Thanks!! on beta 1 and 2 it works fine.

**Comment by heavenstay** [comment]
I encountered the same issue and found that it works on beta 3 when I manually update the plugin locally using the changes from your PR: [flutter/packages#7100](https://github.com/flutter/packages/pull/7100/commits/be43856ab744f9a824d220d2bdeab63f1b078be0).

Thank you for the fix! Just waiting for the PR to be merged to get this resolved in the main branch.

**Comment by eidolonFIRE** [comment]
Confirmed it worked for me.

```
dependency_overrides:
  url_launcher_ios:
    git:
      url: git@github.com:hellohuanlin/packages.git
      path: packages/url_launcher/url_launcher_ios
      ref: urllauncher_fix_main_actor_completion_conformance
```

**Comment by daironigr** [comment]
> Confirmed it worked for me.
> 
> ```
> dependency_overrides:
>   url_launcher_ios:
>     git:
>       url: git@github.com:hellohuanlin/packages.git
>       path: packages/url_launcher/url_launcher_ios
>       ref: urllauncher_fix_main_actor_completion_conformance
> ```

this works perfect!!

**Comment by jmagman** [comment]
I checked beta 4, it's still there:
```swift
    @available(iOS 10.0, *)
    open func open(_ url: URL, options: [UIApplication.OpenExternalURLOptionsKey : Any] = [:], completionHandler completion: (@MainActor @Sendable (Bool) -> Void)? = nil)
```

**Comment by stuartmorgan** [comment]
Since we're seeing a lot of comments on related issues and PRs: if you are experiencing this issue, please run:
`flutter pub upgrade` to update your transitive dependencies (including `url_launcher_ios` which is where this issue and its fix are).
